### PR TITLE
Popcnt optimization

### PR DESCRIPTION
**Pull request by zoecarver**
This patch optimizes count-set-bit functions into a call to LLVM's `popcnt` intrinsic. This is not only good for binary size but, also improves performance. 

To enable the best result, we have to tell llvm the target architecture. If we do this, then (on CPUs that support it) the call to the `popcnt` intrinsic can be optimized to a single `popcnt` instruction. 

Here are some examples (notice, no jumps and no branches):
### Input 1
```swift
func test(_ _x: Int) -> Int {
    var x = _x
    var c = Int()
    while x != 0 {
        c += 1;
        x &= (x - 1);
    }
    return c
}
```

### New output
```
_$s3run4testyS2iF:
	pushq	%rbp
	movq	%rsp, %rbp
	movq	%rdi, %rax
	shrq	%rax
	movabsq	$6148914691236517205, %rcx
	andq	%rax, %rcx
	subq	%rcx, %rdi
	movabsq	$3689348814741910323, %rax
	movq	%rdi, %rcx
	andq	%rax, %rcx
	shrq	$2, %rdi
	andq	%rax, %rdi
	addq	%rcx, %rdi
	movq	%rdi, %rax
	shrq	$4, %rax
	leaq	(%rax,%rdi), %rax
	movabsq	$1085102592571150095, %rcx
	andq	%rax, %rcx
	movabsq	$72340172838076673, %rax
	imulq	%rcx, %rax
	shrq	$56, %rax
	popq	%rbp
	retq
```

### New output (haswell and later, not tested)
```
_$s3run4testyS2iF:
	popcnt  %rax, %rdi
	retq
```

### Old output
```
_$s3run4testyS2iF:
	pushq	%rbp
	movq	%rsp, %rbp
	testq	%rdi, %rdi
	je	LBB1_1
	xorl	%ecx, %ecx
	.p2align	4, 0x90
LBB1_3:
	movq	%rcx, %rax
	incq	%rax
	jo	LBB1_7
	movq	%rdi, %rdx
	decq	%rdx
	jo	LBB1_8
	incq	%rcx
	andq	%rdx, %rdi
	jne	LBB1_3
	popq	%rbp
	retq
LBB1_1:
	xorl	%eax, %eax
	popq	%rbp
	retq
LBB1_7:
	## InlineAsm Start
	## InlineAsm End
	ud2
LBB1_8:
	## InlineAsm Start
	## InlineAsm End
	ud2
```

### Input 2
```swift
func test(_ _x: Int) -> Int {
    var x = _x
    var c = Int()
    var o = 0
    while x != 0 {
        c += 1;
        o += c;
        x &= (x - 1);
    }
    return c - o
}
```

### New output
```
_$s3run4testyS2iF:
	pushq	%rbp
	movq	%rsp, %rbp
	xorl	%eax, %eax
	popq	%rbp
	retq
```

### Old output
```
_$s3run4testyS2iF:
	pushq	%rbp
	movq	%rsp, %rbp
	testq	%rdi, %rdi
	je	LBB1_1
	xorl	%edx, %edx
	xorl	%ecx, %ecx
	.p2align	4, 0x90
LBB1_3:
	movq	%rdx, %rax
	incq	%rax
	jo	LBB1_9
	addq	%rax, %rcx
	jo	LBB1_10
	movq	%rdi, %rsi
	decq	%rsi
	jo	LBB1_11
	incq	%rdx
	andq	%rsi, %rdi
	jne	LBB1_3
	subq	%rcx, %rax
	jo	LBB1_12
LBB1_8:
	popq	%rbp
	retq
LBB1_1:
	xorl	%eax, %eax
	xorl	%ecx, %ecx
	subq	%rcx, %rax
	jno	LBB1_8
LBB1_12:
	## InlineAsm Start
	## InlineAsm End
	ud2
LBB1_10:
	## InlineAsm Start
	## InlineAsm End
	ud2
LBB1_11:
	## InlineAsm Start
	## InlineAsm End
	ud2
LBB1_9:
	## InlineAsm Start
	## InlineAsm End
	ud2
```

This patch is _almost_ ready for @swift-ci to test and benchmark (I'm not sure we'll see a huge improvement, though). I sill need to:

 - [ ] fix the issue with loops
 - [ ] tell llvm the architecture (this should probably be its own patch, though)

### CODE REVIEW

File: include/swift/AST/Builtins.def [hunk 1]
```diff
@@ -462,6 +462,9 @@ BUILTIN_MISC_OPERATION(OnFastPath, "onFastPath", "n", None)
 /// CmpXChg has type (Builtin.RawPointer, T, T) -> (T, Bool).
 BUILTIN_MISC_OPERATION(CmpXChg, "cmpxchg", "", Special)
 
+/// PopCnt has type Builtin.Int64 -> Builtin.Int64.
```
Comment (zoecarver) [inline include/swift/AST/Builtins.def]: Oops, I should make this work for all integer types

**Comment by karwa** [comment]
Just my 2c: we already have FixedWidthInteger.nonZeroBitCount (https://swiftdoc.org/v4.2/protocol/fixedwidthinteger/#var-nonzerobitcount_-int), which is likely a better solution IMO because it’s a language-level fix rather than an optimisation which may or may not happen if your code deviates from the very specific pattern which this recognises.

See this comment thread: https://news.ycombinator.com/item?id=11760293

**Comment by zoecarver** [comment]
Do you mean that the optimization should call out to `nonZeroBitCount`, or that it means we don't need the optimization? If the latter, I think we should keep the optimization because it still helps in cases where people count set bits in another loop. Also, for what it's worth, clang has a similar optimization.

**Comment by jckarter** [comment]
If there's an LLVM-level optimization for this already, then Swift code should already benefit from it.

**Comment by zoecarver** [comment]
There's an intrinsic (which already has a corresponding builtin? I couldn't find it when I made this patch). The intrinsic can be optimized if llvm knows the CPU.

**Comment by jckarter** [comment]
You said that "clang has a similar optimization", which I took to mean turning a popcount loop into the popcnt intrinsic, but I would guess that it's really LLVM applying that optimization. If so, then it ought to already do the same for Swift.

**Comment by jckarter** [comment]
And you can see the same optimization applies in Swift, if you manually avoid the overflow checks:

https://gcc.godbolt.org/z/_kbvlF

So a better approach to enabling this optimization in Swift might be to improve Swift or LLVM's interval analysis to be able to eliminate the overflow checks that will never be triggered in the normal case.

**Comment by jckarter** [comment]
...though `x &= x - 1` does actually need to be `x &= (x &- 1)` to be correct, since otherwise this will crash if `x == Int.min`.

**Comment by zoecarver** [comment]
@jckarter thanks for the links and insight! 

> You said that "clang has a similar optimization", which I took to mean turning a popcount loop into the popcnt intrinsic, but I would guess that it's really LLVM applying that optimization. If so, then it ought to already do the same for Swift.
> 
> You can see that LLVM does the optimization by removing -emit-llvm from the compiler flags here:

I thought clang emitted the unoptimized LLVM IR but it might be optimized. Anyway, if you add the compiler flag `-march=haswell` you can see that the LLVM IR has a call to the popcnt intrinsic.  I took this to mean that clang did the optimization itself (turning a popcount loop into the popcnt intrinsic). 

> And you can see the same optimization applies in Swift, if you manually avoid the overflow checks:

Interesting -- I wish I had seen this before putting this patch together :P Maybe it is actually LLVM doing the optimization. 

If we can guarantee that this will be optimized into a call to the popcnt intrinsic, we could probably safely get rid of the conditional fail. 

How do you think I should proceed on the swift end? I'll see if I can track down the popcnt intrinsic optimization on the LLVM side and then see if I can update it to work with this example too.

**Comment by jckarter** [comment]
The most broadly beneficial place to address this would be in LLVM, adding functionality to determine that the trip count of the loop is limited, so `count += 1` will never overflow, so that the optimizer can eliminate the overflow check and trap and thereby unblock the popcnt optimization. I'm not familiar enough with LLVM to know how best to approach that, though.

**Comment by stephentyrone** [comment]
`.nonzeroBitCount` already lowers directly to the LLVM intrinsic:
```
Sandbox % cat popcnt.swift          
func foo(_ x: Int) -> Int {
  x.nonzeroBitCount
}
Sandbox % swiftc popcnt.swift -O -emit-ir
; ModuleID = '-'
source_filename = "-"
target datalayout = "e-m:o-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-apple-macosx10.15.0"

@__swift_reflection_version = linkonce_odr hidden constant i16 3
@llvm.used = appending global [1 x i8*] [i8* bitcast (i16* @__swift_reflection_version to i8*)], section "llvm.metadata"

; Function Attrs: norecurse nounwind readnone
define i32 @main(i32, i8** nocapture readnone) local_unnamed_addr #0 {
entry:
  ret i32 0
}

; Function Attrs: nounwind readnone
define hidden swiftcc i64 @"$s6popcnt3fooyS2iF"(i64) local_unnamed_addr #1 {
entry:
  %1 = tail call i64 @llvm.ctpop.i64(i64 %0), !range !13
  ret i64 %1
}

; Function Attrs: nounwind readnone speculatable
declare i64 @llvm.ctpop.i64(i64) #2

attributes #0 = { norecurse nounwind readnone "no-frame-pointer-elim"="true" "no-frame-pointer-elim-non-leaf" "target-cpu"="penryn" "target-features"="+cx16,+fxsr,+mmx,+sahf,+sse,+sse2,+sse3,+sse4.1,+ssse3,+x87" }
attributes #1 = { nounwind readnone "no-frame-pointer-elim"="true" "no-frame-pointer-elim-non-leaf" "target-cpu"="penryn" "target-features"="+cx16,+fxsr,+mmx,+sahf,+sse,+sse2,+sse3,+sse4.1,+ssse3,+x87" }
attributes #2 = { nounwind readnone speculatable }

!llvm.module.flags = !{!0, !1, !2, !3, !4, !5, !6, !7, !8}
!swift.module.flags = !{!9}
!llvm.linker.options = !{!10, !11}
!llvm.asan.globals = !{!12}

!0 = !{i32 2, !"SDK Version", [2 x i32] [i32 10, i32 15]}
!1 = !{i32 1, !"Objective-C Version", i32 2}
!2 = !{i32 1, !"Objective-C Image Info Version", i32 0}
!3 = !{i32 1, !"Objective-C Image Info Section", !"__DATA,__objc_imageinfo,regular,no_dead_strip"}
!4 = !{i32 4, !"Objective-C Garbage Collection", i32 83953408}
!5 = !{i32 1, !"Objective-C Class Properties", i32 64}
!6 = !{i32 1, !"wchar_size", i32 4}
!7 = !{i32 7, !"PIC Level", i32 2}
!8 = !{i32 1, !"Swift Version", i32 7}
!9 = !{!"standard-library", i1 false}
!10 = !{!"-lswiftCore"}
!11 = !{!"-lobjc"}
!12 = distinct !{null, null, null, i1 false, i1 true}
!13 = !{i64 0, i64 65}
```
It would be nice to teach the compiler to generate the same optimized code for naive loops, but we really want people to just use the existing Swift operation instead.

**Comment by stephentyrone** [comment]
I don't *think* there's any need for changes at the Swift level to support this; this should probably be purely an LLVM optimization (but I also haven't thought about it in that much detail yet).

**Comment by stephentyrone** [comment]
This all *does* raise the vaguely interesting question of where exactly the ctpop Builtin is currently getting defined, given that you thought you had to add it and a simple search doesn't turn it up, but somehow the stdlib is able to use it from the Builtin module anyway:
```
swift % grep -ri ctpop .
./validation-test/stdlib/Prototypes/PersistentVector.swift.gyb:    return Int(Builtin.int_ctpop_Int32(_bits._value))
./validation-test/stdlib/Prototypes/PersistentVector.swift.gyb:    return Int(Builtin.int_ctpop_Int64(_bits._value))
./validation-test/stdlib/Prototypes/PersistentVector.swift.gyb:    return Int(Int32(Builtin.int_ctpop_Int32(_bits._value)))
./stdlib/public/core/IntegerTypes.swift.gyb:        Builtin.int_ctpop_Int${bits}(self._value)
```
I'm sure there's a simple explanation, but it's late here and I'm not seeing it.

**Comment by zoecarver** [comment]
Yes, I was also curious about that. The built `IntegerTypes.swift` file has references to the builtin, but I cannot find the builtin defined anywhere. Is it possible that it is just falling back to indexing an llvm intrinsic?

**Comment by jckarter** [comment]
`Builtin.int_<intrinsic_name>` automatically references an LLVM intrinsic by name.

**Comment by zoecarver** [comment]
@jckarter good to know, thanks :)

**Comment by zoecarver** [comment]
As discussed, this should be an llvm change, not a swift change. Closing.

**Pull request closed without merging**